include/seqan3/core/detail/strong_type.hpp

```diff
@@ -0,0 +1,546 @@
+ * ```
+ * Our interface could now be changed to:
+ *
+ * ```cpp
+ * template <forward_range_concept fwd_rng_type>
+ * bool search(fwd_rng_type & rng, window_size const w, error const e)
+ * {
+ *  ...
+ * };
+ * ```
+ *
+ * Now the user is forced to pass the parameters as their named type. If the parameter order is mixed up by accident
+ * the compiler would emit an error message, since the `error` type is not convertible to the `window_size` type and
+ * vice versa.
+ *
+ * ### Adding Skills
+ *
+ * In most cases it is sufficient to protect the user interface from misuse by using strong types. Within the
+ * implementation the underlying value can then be extracted using the `getter`-member functions of the
+ * seqan3::detail::strong_type class. However, there might be scenarios, where the strong type is exposed to the user
+ * to work with, but with a restricted set of operations that can be applied to the type.
+ * A familiar use case are the time typedefs of the [std::chrono](http://en.cppreference.com/w/cpp/header/chrono)
+ * library. It is convenient for the user to subtract two time values representing, for example seconds, to get the
+ * duration between two time points. But not all operations like modulo or multiplication are really useful for this.
+ * In order to add skills to the seqan3::detail::strong_type the typedef can be further specialized with
+ * operations from the seqan::detail::skill enum. The corresponding operators will be added if the underlying type
+ * supports these operations as well.
+ * For example, we could further specify our error type to support increment and decrement operations.
+ *
+ * ```cpp
+ * using error = detail::strong_type<unsigned, struct search_error_tag, detail::skill::increment | detail::skill::decrement>;
+ *
+ * error e{4};
+ * --e;
+ * ++e;
+ * ```
+ */
+template <typename unnamed_type, typename name_tag, skill skills = skill::none>
+class strong_type : public skills_base<strong_type<unnamed_type, name_tag, skills>, skills>
+{
+public:
+
+    /*!\name Constructor, destructor and assignment.
+     * \brief The standard functions are explicitly defaulted.
+     * \{
+     */
+    constexpr strong_type() = default;
+    constexpr strong_type(strong_type const &) = default;
+    constexpr strong_type(strong_type &&) = default;
+    constexpr strong_type & operator= (strong_type const &) = default;
+    constexpr strong_type & operator= (strong_type &&) = default;
+    ~strong_type() = default;
+
+    //!|brief Construction from underlying value type.
+    constexpr explicit strong_type(unnamed_type const & _value) : value(_value) /*noexcept(noexcept(value(_value)))*/
+    {}
+
+    //!|brief Move construction from underlying value type.
+    template<typename unnamed_type_ = unnamed_type>
+    //!\cond
+        requires !std::is_lvalue_reference_v<unnamed_type_>
+    //!\endcond
+    constexpr explicit strong_type(unnamed_type_ && _value) : value(std::move(_value))
+    {}
```

Comment:
one interface should be enough:
```cpp
constexpr explicit strong_type(unnamed_type _value) : value(std::move(_value))
{}
```
or not?

Reply:
Let's say, I am at the moment not aware of a point, where it might be required, but your solution wouldn't work if the strong type should be created for an l-value reference.
In this case: value has the type `type &`. And I cannot bind an rvalue of `type` to an non-const l-value of `type`. I could lave the move away, but then it induces two copies of `_value` in most use cases.

Reply:
> your solution wouldn't work if the strong type should be created for an l-value reference.

lvalue-references are currently not supported anyway as per your member definition.  Also I can imagine no scenario where they would be desirable...

Reply:
Well, yes, if the unnamed_type is declared e.g., `std::string &`

Reply:
it can't be because you have a member variable of `unnamed_type` that is not being initialised in default initialisation.

Reply:
I don't follow.
Clearly, the default c'tor will get deleted, but then there is a compiler error if someone tries to use the default c'tor in this context. So it does work as expected (and is tested)

Reply:
But I suggest another solution. I will drop the support for the l-value reference in favour of using std::reference_wrapper to add functionality for this scenario.

Reply:
Just to get it straight, one other aspect of this point, is the number of moves.
In your version:

```cpp
using bar = strong_type<foo, struct bar_tag>;
foo f;
bar b{std::move(f)};
```
Would induce two moves. In the old interface it would be still only one move construction executed.